File: app/src/main/java/ru/geekbrains/notes/DatepickerFragment.java
package ru.geekbrains.notes;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.DatePicker;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.text.DateFormat;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

import ru.geekbrains.notes.domain.note.Note;
import ru.geekbrains.notes.ui.SharedPref;

import static ru.geekbrains.notes.Constant.MILISECOND;


public class DatepickerFragment extends Fragment {

    private static final String ARG = "NOTE_ID";

    public DatepickerFragment() {
        // Required empty public constructor
    }

    public static DatepickerFragment newInstance(Note note) {
        Log.v("Debug1", "DatepickerFragment newInstance");
        DatepickerFragment fragment = new DatepickerFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(ARG, note);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        Log.v("Debug1", "DatepickerFragment onCreateView");

        View v = inflater.inflate(R.layout.fragment_datepicker, container, false);
        DatePicker datePicker = v.findViewById(R.id.datepicker);

        if (getArguments() != null) {
            Note note;
            note = getArguments().getParcelable(ARG);
            long date = note.getDate() * MILISECOND;
            Calendar calendar = Calendar.getInstance();
            calendar.setTimeInMillis(date);
            int year = calendar.get(Calendar.YEAR);
            int month = calendar.get(Calendar.MONTH);
            int day = calendar.get(Calendar.DAY_OF_MONTH);

            datePicker.init(year, month, day, (view, year1, monthOfYear, dayOfMonth) -> {

                Log.v("Debug1", "DatepickerFragment onDateChanged into");

                Note note1 = getArguments().getParcelable(ARG);
                Calendar calendar1 = Calendar.getInstance();
                calendar1.set(year1, monthOfYear, dayOfMonth);
                long newDate = calendar1.getTimeInMillis() / 1000;

                if (getActivity() != null) {
                    List<Note> notes = ((MyApplication) getActivity().getApplication()).getNotes();
                    note1.setDate(newDate);
                    notes.set(note1.getID(), note1);
                    ((MyApplication) getActivity().getApplication()).setNotes(notes);

                    new SharedPref(getActivity()).saveNotes(notes);

                    //getFragmentManager().beginTransaction().remove(DatepickerFragment.this).commit();

                    if (getFragmentManager() != null) {
                        FragmentTransaction fragmentTransaction = getFragmentManager().beginTransaction();
                        fragmentTransaction.remove(DatepickerFragment.this);
                        fragmentTransaction.commit();

                        LinearLayout linearLayout = getActivity().findViewById(R.id.linearLayoutIntoScrollViewCont);
                        TextView textViewTop = linearLayout.findViewWithTag(note.getID());

                        Log.v("Debug1", "DatepickerFragment onDateChanged into note.getID()=" + note.getID());

                        DateFormat f = DateFormat.getDateTimeInstance(DateFormat.DEFAULT, DateFormat.DEFAULT, Locale.getDefault());
                        String dateStr = f.format(newDate * 1000);

                        textViewTop.setText(dateStr);
                    }
                }
            });
        }
        return v;
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Log.v("Debug1", "DatepickerFragment onViewCreated");

        if (getArguments() != null) {
            Note note;
            note = getArguments().getParcelable(ARG);
            long date = note.getDate() / MILISECOND;
            Calendar calendar = Calendar.getInstance();
            calendar.setTimeInMillis(date);
        }
    }
}

File: app/src/main/java/ru/geekbrains/notes/ui/DatepickerFragment.kt
package ru.geekbrains.notes.ui

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.DatePicker
import androidx.fragment.app.Fragment
import ru.geekbrains.notes.Constant.MILISECOND
import ru.geekbrains.notes.GlobalVariables
import ru.geekbrains.notes.R
import ru.geekbrains.notes.SharedPref
import ru.geekbrains.notes.observer.Publisher
import ru.geekbrains.notes.observer.PublisherHolder
import java.util.*

class DatepickerFragment : Fragment() {
    private var noteId = 0
    private var publisher: Publisher? = null
    private var yearFromDP = 0
    private var monthOfYearFromDP = 0
    private var dayOfMonthFromDP = 0
    @SuppressLint("UseRequireInsteadOfGet")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.fragment_datepicker, container, false)
        if (arguments != null && activity != null) {
            noteId = requireArguments().getInt(ARG, 0)
            Log.v("Debug1", "DatepickerFragment onCreateView noteId=$noteId")
            val note = (requireActivity().application as GlobalVariables).getNoteByNoteId(noteId)
            val date: Long = note.dateEdit * MILISECOND
            val calendar = Calendar.getInstance()
            calendar.timeInMillis = date
            val year = calendar[Calendar.YEAR]
            val month = calendar[Calendar.MONTH]
            val day = calendar[Calendar.DAY_OF_MONTH]
            val buttonOk = v.findViewById<Button>(R.id.button_dp_ok)
            buttonOk.setOnClickListener {
                Log.v("Debug1", "DatepickerFragment button_ok")
                val calendarNew = Calendar.getInstance()
                calendarNew[yearFromDP, monthOfYearFromDP] = dayOfMonthFromDP
                val newDate: Long = calendarNew.timeInMillis / MILISECOND
                if (this@DatepickerFragment.activity != null) {
                    note.dateEdit = newDate
                    (this@DatepickerFragment.activity!!.application as GlobalVariables).setNoteById(noteId, note)
                    val notes = (this@DatepickerFragment.activity!!.application as GlobalVariables).getNotes()
                    SharedPref(this@DatepickerFragment.activity!!).saveNotes(notes!!)
                    if (publisher != null) {
                        publisher!!.notify(noteId, 3)
                    }
                    if (this@DatepickerFragment.activity != null) {
                        val fragmentManager = this@DatepickerFragment.activity!!.supportFragmentManager
                        fragmentManager.popBackStack()
                    }
                }
            }
            val buttonCancel = v.findViewById<Button>(R.id.button_dp_cancel)
            buttonCancel.setOnClickListener {
                if (this@DatepickerFragment.activity != null) {
                    val fragmentManager = this@DatepickerFragment.activity!!.supportFragmentManager
                    fragmentManager.popBackStack()
                }
            }
            val datePicker = v.findViewById<DatePicker>(R.id.datepicker)
            datePicker.init(year, month, day) { _: DatePicker?, year1: Int, monthOfYear: Int, dayOfMonth: Int ->
                Log.v("Debug1", "DatepickerFragment onDateChanged into")
                yearFromDP = year1
                monthOfYearFromDP = monthOfYear
                dayOfMonthFromDP = dayOfMonth
            }
        }
        return v
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Log.v("Debug1", "DatepickerFragment onViewCreated")
        if (arguments != null && activity != null) {
            noteId = requireArguments().getInt(ARG, 0)
            Log.v("Debug1", "DatepickerFragment onViewCreated noteId=$noteId")
            val note = (requireActivity().application as GlobalVariables).getNoteByNoteId(noteId)
            val date: Long = note.dateEdit / MILISECOND
            val calendar = Calendar.getInstance()
            calendar.timeInMillis = date
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        MainActivity.setTitle(activity, "Правка даты")
        Log.v("Debug1", "DatepickerFragment onAttach")
        if (context is PublisherHolder) {
            publisher = (context as PublisherHolder).publisher
        }
    }

    override fun onDetach() {
        super.onDetach()
        Log.v("Debug1", "DatepickerFragment onDetach")
        publisher = null
    }

    companion object {
        private const val ARG = "NOTE_ID"
        const val TAG = "DatepickerFragment"
        @JvmStatic
        fun newInstance(noteId: Int): DatepickerFragment {
            Log.v("Debug1", "DatepickerFragment newInstance noteId=$noteId")
            val fragment = DatepickerFragment()
            val bundle = Bundle()
            bundle.putInt(ARG, noteId)
            fragment.arguments = bundle
            return fragment
        }
    }
}